lib/ContentTypeParser.js

```diff
@@ -1,67 +1,181 @@
+/* eslint-disable no-useless-return */
 'use strict'
 
+const lru = require('tiny-lru')
+
 function ContentTypeParser () {
   this.customParsers = {}
-  this.parserList = []
+  this.customParsers['application/json'] = defaultJsonParser
+  this.parserList = ['application/json']
+  this.cache = lru(100)
 }
 
-ContentTypeParser.prototype.add = function (contentType, fn) {
+ContentTypeParser.prototype.add = function (contentType, opts, parser) {
   if (typeof contentType !== 'string') throw new Error('The content type should be a string')
   if (contentType.length === 0) throw new Error('The content type cannot be an empty string')
-  if (typeof fn !== 'function') throw new Error('The content type handler should be a function')
+  if (typeof parser !== 'function') throw new Error('The content type handler should be a function')
 
   if (this.hasParser(contentType)) {
     throw new Error(`Content type parser '${contentType}' already present.`)
   }
+
+  parser.asString = opts.asString === true
+  parser.asBuffer = opts.asBuffer === true
+
   if (contentType === '*') {
     this.parserList.push('')
-    this.customParsers[''] = fn
+    this.customParsers[''] = parser
   } else {
-    this.parserList.unshift(contentType)
-    this.customParsers[contentType] = fn
+    if (contentType !== 'application/json') {
+      this.parserList.unshift(contentType)
+    }
+    this.customParsers[contentType] = parser
   }
 }
 
 ContentTypeParser.prototype.hasParser = function (contentType) {
+  if (contentType === 'application/json') return false
```

Comment:
What about after a user sets a custom JSON parser? I think this would handle that case:

```js
if (contentType === 'application/json') {
  return this.customParser['application/json'] !== defaultJsonParser
}
```